### PR TITLE
More elaborate exceptions.

### PR DESCRIPTION
**Pull request by Hixie**
The elaboration will continue until morale improves.

Specific exceptions considered here:
- size setter checking when you set the size
- layout verifying that you do set the size
- hitTest verifying that you aren't dirty and have a size
- flex complaining about canFlex

### CODE REVIEW

**Comment by Hixie** [comment]
Fixes https://github.com/flutter/flutter/issues/2555.

File: packages/flutter/lib/src/rendering/box.dart [hunk 1]
```diff
@@ -520,8 +520,36 @@ abstract class RenderBox extends RenderObject {
...
+    assert(sizedByParent || !debugDoingThisResize);
+    assert(() {
+      if ((sizedByParent && debugDoingThisResize) ||
+          (!sizedByParent && debugDoingThisLayout))
+        return true;
+      assert(!debugDoingThisResize);
+      String contract, violation, hint;
+      if (debugDoingThisLayout) {
+        assert(sizedByParent);
+        violation = 'It appears that the size setter was called from performLayout().';
```
Comment (abarth) [inline packages/flutter/lib/src/rendering/box.dart]: Do you want to say something about sizedByParent here?

File: packages/flutter/lib/src/rendering/box.dart [hunk 2]
```diff
@@ -520,8 +520,36 @@ abstract class RenderBox extends RenderObject {
...
+        hint = '';
+      } else {
+        violation = 'The size setter was called from outside layout (neither performResize() nor performLayout() were being run for this object).';
+        if (owner != null && owner.debugDoingLayout)
+          hint = 'Only the object itself can set its size. It is a contract violation for other objects to set it.';
+      }
+      if (sizedByParent)
+        contract = 'Since this RenderBox has sizedByParent set to true, it must set its size in performResize().';
```
Comment (abarth) [inline packages/flutter/lib/src/rendering/box.dart]: s/Since/Because/

File: packages/flutter/lib/src/rendering/box.dart [hunk 3]
```diff
@@ -520,8 +520,36 @@ abstract class RenderBox extends RenderObject {
...
+      else
+        contract = 'Since this RenderBox has sizedByParent set to false, it must set its size in performLayout().';
```
Comment (abarth) [inline packages/flutter/lib/src/rendering/box.dart]: ditto

**Comment by abarth** [comment]
LGTM

**Pull request merged**